src/AWSS3.jl

```diff
@@ -195,19 +195,58 @@ end
 
 s3_get_meta(a...; b...) = s3_get_meta(global_aws_config(), a...; b...)
 
+function _s3_exists_file(aws::AbstractAWSConfig, bucket, path)
+    q = Dict("prefix" => path, "delimiter" => "", "max-keys" => 1)
+    l = S3.list_objects_v2(bucket, q; aws_config=aws)
+    c = get(l, "Contents", nothing)
+    c ≡ nothing && return false
+    return get(c, "Key", "") == path
+end
+
 """
-    s3_exists([::AbstractAWSConfig], bucket, path [version=], kwargs...)
+    _s3_exists_dir(aws::AbstractAWSConfig, bucket, path)
+
+Internal, called by [`s3_exists`](@ref).
+
+Checks whether the given directory exists.  This is a bit subtle because of how the
+AWS API handles empty directories.  Empty directories are really just 0-byte nodes
+which are named like directories, i.e. their name has a trailing `"/"`.
 
-Is there an object in `bucket` at `path`?
+What this function does is, given a directory name `dir/`, check for all keys which
+are lexographically greater than `dir.`.  The reason for this is that, if `dir/`
+is a 0-byte node, checking for it directly will not reveal its existence due to
+some rather peculiar design choices on the part of the S3 developers.
+
+In all such cases, if the directory exists it will be seen in the *first* item
+returned from `S3.list_objects_v2`: for empty directories this is because using
+`start-after` explicitly excludes `dir.` and `dir/` is next, for directories
+with actual keys, it is guaranteed that the first contained key will start with
+the directory name.
```

Comment:
maybe this is slightly clearer:
```suggestion
returned from `S3.list_objects_v2`: for empty directories this is because using
`start-after` explicitly excludes `dir.` itself and `dir/` is next; for directories
with actual keys, it is guaranteed that the first contained key will start with
the directory name.
```